src/main/java/hudson/security/UserAttributesHelper.java

```diff
@@ -0,0 +1,285 @@
+ * the UserDetails that will be used by the SecurityRealm
+ */
+@Restricted(NoExternalUse.class)
+public class UserAttributesHelper {
+    private static final Logger LOGGER = LDAPSecurityRealm.LOGGER;
+    // https://support.microsoft.com/en-us/help/305144/how-to-use-the-useraccountcontrol-flags-to-manipulate-user-account-pro
+    // https://docs.microsoft.com/en-us/windows/win32/adschema/a-useraccountcontrol
+    private static final String ATTR_USER_ACCOUNT_CONTROL = "userAccountControl";
+    // https://docs.microsoft.com/en-us/windows/win32/adschema/a-accountexpires
+    private static final String ATTR_ACCOUNT_EXPIRES = "accountExpires";
+    // https://ldapwiki.com/wiki/Draft-behera-ldap-password-policy
+    private static final String ATTR_LOGIN_DISABLED = "loginDisabled";
+    private static final String ATTR_ORACLE_IS_ENABLED = "orclIsEnabled";
+    private static final String ATTR_PWD_ACCOUNT_LOCKED_TIME = "pwdAccountLockedTime";
+    private static final String ATTR_PWD_START_TIME = "pwdStartTime";
+    private static final String ATTR_PWD_END_TIME = "pwdEndTime";
+    private static final String ATTR_LOGIN_EXPIRATION_TIME = "loginExpirationTime";
+    private static final String ATTR_PWD_LOCKOUT = "pwdLockout";
+    // https://ldapwiki.com/wiki/Locked%20By%20Intruder
+    private static final String ATTR_LOCKED_BY_INTRUDER = "lockedByIntruder";
+    // for Windows Server 2003-based domain
+    // https://docs.microsoft.com/en-us/windows/win32/adschema/a-msds-user-account-control-computed
+    private static final String ATTR_USER_ACCOUNT_CONTROL_COMPUTED = "msDS-User-Account-Control-Computed";
+    // for ADAM (Active Directory Application Mode), replace the ADS_UF_DISABLED
+    // https://docs.microsoft.com/en-us/windows/win32/adschema/a-msds-useraccountdisabled
+    private static final String ATTR_USER_ACCOUNT_DISABLED = "msDS-UserAccountDisabled";
+    // for ADAM, replace the ADS_UF_PASSWORD_EXPIRED
+    // https://docs.microsoft.com/en-us/windows/win32/adschema/a-msds-userpasswordexpired
+    private static final String ATTR_USER_PASSWORD_EXPIRED = "msDS-UserPasswordExpired";
+    private static final String ACCOUNT_DISABLED = "000001010000Z"; // other GeneralizedTime values indicate account is locked as of that time
+
+    // https://docs.microsoft.com/en-us/windows/desktop/adschema/a-accountexpires
+    // constant names follow the code in Iads.h
+    private static final long ACCOUNT_NO_EXPIRATION = 0x7FFF_FFFF_FFFF_FFFFL;
+    private static final int ADS_UF_DISABLED = 0x0002;
+    private static final int ADS_UF_LOCK_OUT = 0x0010;
+    private static final int ADS_DONT_EXPIRE_PASSWORD = 0x1_0000;
+    private static final int ADS_UF_PASSWORD_EXPIRED = 0x80_0000;
+
+    private static final ChronoFormatter<Moment> GENERALIZED_TIME_FORMATTER;
+
+    static {
+        // https://stackoverflow.com/a/42383153
+        // this static block licensed https://creativecommons.org/licenses/by-sa/3.0/
```

Comment:
Note to self: since this code snippet is licensed differently, it might be a good idea to split into its own file with the proper license header.